rasa/server.py

```diff
@@ -779,14 +780,25 @@ def _get_events_from_request_body(request: Request) -> List[Event]:
             with app.active_training_processes.get_lock():
                 app.active_training_processes.value += 1
 
-            model_path = await train_async(
+            info = dict(
                 domain=domain_path,
                 config=config_path,
                 training_files=temp_dir,
-                output_path=model_output_directory,
+                output=model_output_directory,
                 force_training=rjs.get("force", False),
             )
 
+            loop = asyncio.get_event_loop()
+
+            from rasa import train as train_model
+
+            # Declare `model_path` upfront to avoid pytype `name-error`
+            model_path: Optional[Text] = None
+            # pass `None` to run in default executor
+            model_path = await loop.run_in_executor(
+                None, functools.partial(train_model, **info)
```

Comment:
So here we've chosen to run the training function on a separate thread instead of a separate process, right? Is there a chance the training thread might use the GIL in a way that might block Sanic from processing requests? (If Tensorflow does not lock the GIL significantly then I guess it's not a problem)

Reply:
> Is there a chance the training thread might use the GIL in a way that might block Sanic from processing requests? 

That shouldn't matter. When running things in a separate thread Python will switch between the threads every x milliseconds so that other threads can continue processing.

Reply:
5 ms btw https://stackoverflow.com/a/37911403/3429596